Require the user password when generating backup codes.

xknown opened the pull request:
#### Changes proposed in this Pull Request

* Requires the current user password when generating new backup codes.

#### Testing instructions

User with 2fa enabled:
* Navigate to `/me/security/two-step`
* Click on the `Generate new backup codes` button, you should notice a password prompt. <img width="1044" alt="Screenshot 2021-11-08 at 15 50 31" src="https://user-images.githubusercontent.com/1081870/140763743-0fb5590b-77e5-4ef0-beca-418f70f4c3fb.png">
* Backup codes are generated as always (this depends on the following backend change D69484)

User without 2fa enabled.
* Navigate to `/me/security/two-step`
* Enable 2fa with sms and/or with an app
* Backup codes are shown without the need to enter the user password.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

xknown committed:
Require the user password when generating backup codes.

This also removes the extra request done when setting up 2fa on a WP.com
account. The validate endpoint returns already a list of backup codes
if the provided code is valid.

xknown committed:
Remove unneeded flag.

a8ci18n commented on client/me/security-2fa-backup-codes/password-prompt.jsx:
```diff
@@ -0,0 +1,67 @@
+						type="password"
+						ref={ ( input ) => ( this.passwordInput = input ) }
+						value={ this.state.userPassword }
+						placeholder={ this.props.translate( 'Your WordPress.com password' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **53** times:
`translate( 'WordPress.com Password' )` *ES Score: 6*

xknown committed:
Notify the user if there's an error when generating backup codes.

xknown committed:
Pass the appropriate argument to the backend

xknown committed:
Disable the button while we are validating the password

xknown committed:
Use the i18n suggestion

xknown committed:
Update the password field name

a8ci18n commented on client/me/security-2fa-backup-codes/index.jsx:
```diff
@@ -96,6 +116,34 @@ class Security2faBackupCodes extends Component {
+				lastError?.error === 'authentication_failed'
+					? this.props.translate( 'Invalid password' )
+					: this.props.translate(
+							'Unable to generate backup codes right now. Please try again later.'
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **29** times:
`translate( 'Unable to validate codes right now. Please try again later.' )` *ES Score: 7*

tyxla replied:
Could we reuse the existing string? Seems like a good idea to me.

xknown replied:
Not really ideal in this context I think. This error message will be shown when the request to generate new codes fails, and we aren't really validating codes :) (currently there's no error handling if the API call fails)

tyxla replied:
Oops, I quickly skimmed through it and thought the only difference was the absence of "backup" there.

Disregard 🤦

xknown replied:
No worries, thanks for the review :)

a8ci18n commented on client/me/security-2fa-backup-codes/password-prompt.jsx:
```diff
@@ -0,0 +1,68 @@
+						type="password"
+						ref={ ( input ) => ( this.passwordInput = input ) }
+						value={ this.state.userPassword }
+						placeholder={ this.props.translate( 'WordPress.com password' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **53** times:
`translate( 'WordPress.com Password' )` *ES Score: 10*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/require-user-password-2fa-backup-codes)

tyxla replied:
We can definitely reuse that string

tyxla commented on client/lib/two-step-authorization/index.js:
```diff
@@ -200,8 +200,8 @@ TwoStepAuthorization.prototype.sendSMSCode = function ( callback ) {
-TwoStepAuthorization.prototype.backupCodes = function ( callback ) {
-	wpcom.me().backupCodes( ( error, data ) => {
+TwoStepAuthorization.prototype.backupCodes = function ( userPassword, callback ) {
+	wpcom.me().backupCodes( userPassword, ( error, data ) => {
```
While here, have you considered refactoring this to use `wpcom.req.post()` directly instead of proxying it through a separate `UndocumentedMe` method? We've been trying to remove the monolithic `undocumented` methods and we could really use some help in those areas.

xknown replied:
Update the code in https://github.com/Automattic/wp-calypso/pull/57560/commits/6c28f88dc6f5d7cbc6ff7a8e76ace49f7b08bc14, and also decided to remove the `backupCodes` method from the `undocumented` library -- I made sure there are no other calls elsewhere.

tyxla commented on client/me/security-2fa-backup-codes/index.jsx:
```diff
@@ -24,29 +25,48 @@ class Security2faBackupCodes extends Component {
+
+	toggleBackupCodePassword = ( event ) => {
+		event.preventDefault();
+		this.setState( {
+			addingPassword: ! this.state.addingPassword,
+			generatingCodes: false,
+			lastError: null,
+			showPrompt: true,
+		} );
```
This can potentially be using an inconsistent state if multiple updates are batched together (you can read [this](https://reactjs.org/docs/react-component.html#setstate) and particularly [this](https://stackoverflow.com/questions/48563650/does-react-keep-the-order-for-state-updates/48610973#48610973) to understand why). 

To get around this, we use the callback form of `setState()`, and that way we ensure that we're using the up-to-date previous state:

```suggestion
		this.setState( ( prevState ) => ( {
			addingPassword: ! prevState.addingPassword,
			generatingCodes: false,
			lastError: null,
			showPrompt: true,
		} ) );
```

tyxla commented on client/me/security-2fa-backup-codes/index.jsx:
```diff
@@ -96,6 +116,34 @@ class Security2faBackupCodes extends Component {
+				<Notice
+					status="is-error"
+					text={ friendlyMessage }
+					onDismissClick={ () => this.setState( { lastError: null } ) }
```
I think it's a good idea to move this event handler to a separate method in the class, so it won't be redeclared on every re-render of this notice.

tyxla commented on client/me/security-2fa-backup-codes/password-prompt.jsx:
```diff
@@ -0,0 +1,68 @@
+	};
+
+	componentDidMount = () => {
+		this.passwordInput.focus();
```
Should we be using the `autoFocus` prop of `FormTextInput` instead? 

In any case, if we can avoid auto-focusing it would be best. Or if we really need to auto focus the field, let's do it and use `autoFocus` and intentionally disable the `jsx-a11y/no-autofocus` ESLint rule on that line. That can help if we ever decide to do a pass on improving the state of all those accessibility exceptions.

xknown replied:
Updated the code in https://github.com/Automattic/wp-calypso/pull/57560/commits/3dbb7178c35719161ae496a2d9d75a0603168d47

tyxla commented on client/me/security-2fa-backup-codes/password-prompt.jsx:
```diff
@@ -0,0 +1,68 @@
+import FormSettingExplanation from 'calypso/components/forms/form-setting-explanation';
+import FormTextInput from 'calypso/components/forms/form-text-input';
+
+class Security2faBackupCodesPasswordPromt extends React.Component {
```
This looks like a typo:

```suggestion
class Security2faBackupCodesPasswordPrompt extends React.Component {
```

tyxla commented on client/me/security-2fa-backup-codes/password-prompt.jsx:
```diff
@@ -0,0 +1,68 @@
+	}
+}
+
+export default localize( Security2faBackupCodesPasswordPromt );
```
```suggestion
export default localize( Security2faBackupCodesPasswordPrompt );
```

tyxla commented on client/lib/wpcom-undocumented/lib/me.js:
```diff
@@ -98,10 +98,11 @@ UndocumentedMe.prototype.setPeerReferralLinkEnable = function ( enable, callback
 	const args = {
 		apiVersion: '1.1',
 		path: '/me/two-step/backup-codes/new',
+		body: { password },
```
Any security concerns with the fact that the user password is going to that endpoint in plain text? 

In general, is it concerning that we're adding another place where we're expecting the user password, which could potentially be another attack vector? Or is that just fine, considering that the endpoint would require the user to be authenticated in the first place?

xknown replied:
What sort of attack vectors do you think might be possible if we use this endpoint?

An alternative would be to send this to `wp-login`, but not sure if this would be really different to what we are doing here. To get a valid result on this endpoint, the request needs to be authenticated (with a "fresh" 2fa cookie). We'll only add a password validation step along with some throttling to limit how many times one could test the password.

tyxla replied:
I didn't mean anything specific, and I don't have any specific concerns. I just wanted to get your point of view as someone who has a better understanding of security than me 😉

xknown replied:
Thanks for the question :) I had a somewhat similar concern (passing around the user password in different parts of code) when I was making the change, and I was curious if you have something particular in mind.

tyxla replied:
FWIW [we are doing the same](https://github.com/Automattic/wp-calypso/blob/migrate/undocumented-publicize/client/state/login/actions/login-user.js#L34) when logging in the user already, so I think we should be good.

jsnajdr replied:
GitHub web UI has something similar: before entering some sensitive parts of the site, you need to reenter your password, and it leads to a `post( '/sessions/sudo', { sudo_password } )` request.

The only concern is that there's now one more REST endpoint that handles passwords, and one more thing to audit and be careful about.

But as long the request is a `POST` and is encrypted, it's technically OK.

xknown replied:
Do you maybe know what other endpoints handle passwords? The only one I know is the settings endpoint to update the current user password. I agree however that we might probably consider having a similar feature to harden some areas of the application.

On an unrelated note, do you maybe know why the e2e tests might be failing. This change doesn't touch any dependency.

tyxla replied:
> Do you maybe know what other endpoints handle passwords? The only one I know is the settings endpoint to update the current user password. I agree however that we might probably consider having a similar feature to harden some areas of the application.

We directly use the canonical backend WP.com login as an endpoint when logging in - see https://github.com/Automattic/wp-calypso/blob/f2e6f88ae65c8571bbac21b284c29b48ec987598/client/state/login/utils.js#L165.

> On an unrelated note, do you maybe know why the e2e tests might be failing. This change doesn't touch any dependency.

It seems the docker build failed for this branch, but a rebase of the branch should fix it, it doesn't seem related.

tyxla commented on client/me/security-2fa-backup-codes/password-prompt.jsx:
```diff
@@ -0,0 +1,68 @@
+import { localize } from 'i18n-calypso';
+import PropTypes from 'prop-types';
+import React from 'react';
```
We are only using `Component`, so perhaps let's just import that instead of all of `React`.

tyxla commented on client/me/security-2fa-backup-codes/password-prompt.jsx:
```diff
@@ -0,0 +1,68 @@
+class Security2faBackupCodesPasswordPromt extends React.Component {
+	static propTypes = {
+		onSubmit: PropTypes.func.isRequired,
+		onCancel: PropTypes.func.isRequired,
```
Uber nit: Should we sort those alphabetically?

tyxla commented on client/me/security-2fa-backup-codes/password-prompt.jsx:
```diff
@@ -0,0 +1,68 @@
+						id="password"
+						name="password"
+						type="password"
+						ref={ ( input ) => ( this.passwordInput = input ) }
```
Using the `autoFocus` prop would also help us remove the `ref`.

tyxla commented on client/me/security-2fa-backup-codes/password-prompt.jsx:
```diff
@@ -0,0 +1,68 @@
+						) }
+					</FormSettingExplanation>
+				</FormFieldset>
+				<FormButton disabled={ 0 === this.state.userPassword.length || this.props.isDisabled }>
```
Nit: You can also just negate it for brevity:

```suggestion
				<FormButton disabled={ ! this.state.userPassword.length || this.props.isDisabled }>
```

tyxla commented on client/me/security-2fa-backup-codes/password-prompt.jsx:
```diff
@@ -0,0 +1,68 @@
+					/>
+					<FormSettingExplanation>
+						{ this.props.translate(
+							'Your WordPress.com password is required to generate new backup codes.'
```
I think a screenshot or two in the PR description would be helpful for the translators.

tyxla commented on client/me/security-2fa-setup-backup-codes/index.jsx:
```diff
@@ -67,13 +41,14 @@ class Security2faSetupBackupCodes extends Component {
-		if ( this.state.lastError ) {
-			return null;
+		const { backupCodes } = this.props;
+		if ( ! backupCodes || backupCodes.length === 0 ) {
```
Suggestion for brevity:

```suggestion
		if ( ! backupCodes || ! backupCodes.length ) {
```

tyxla reviewed (commented):
Nice work with this one! I think it's looking well in general! 

I've left some feedback, let me know if you need anything in the meantime.

Happy to take another look once feedback is addressed.

xknown committed:
Update client/me/security-2fa-setup-backup-codes/index.jsx

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

xknown committed:
Update client/me/security-2fa-backup-codes/password-prompt.jsx

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

xknown committed:
Address received feedback

- Use autoFocus.
- Import only Component instead of React
- Use the suggested i18n string

xknown committed:
Fix typo

xknown committed:
Add a onDismissClick method instead of re-declaring it on render

xknown committed:
Use `setState(fn)` to avoid potential issues when state updates are
batch updated.

See https://github.com/Automattic/wp-calypso/pull/57560#discussion_r744707296

xknown committed:
`wpcom.undocumented()` is deprecated.

Use `wpcom.req.post()` directly.

jsnajdr commented on client/me/security-2fa-backup-codes/password-prompt.jsx:
```diff
@@ -0,0 +1,65 @@
+				</FormButton>
+				<FormButton onClick={ this.props.onCancel } isPrimary={ false }>
+					{ this.props.translate( 'Cancel' ) }
+				</FormButton>
```
I assume the first button (Generate) has type `submit` (the default) and submits the form? Then the "Cancel" button should have a `type="button"` prop so that it doesn't submit the form. That's cleaner than calling `event.preventDefault()` in the click handler, which now must be there to prevent the submission.

xknown replied:
Makes sense, updated the code in https://github.com/Automattic/wp-calypso/pull/57560/commits/abc0430c6880412d8c667a3311d96e950f889307

xknown committed:
Use `type="button"` to avoid the unnecessary `event.preventDefault()`
call.

tyxla commented:
Oops, something happened with the branch, looks like a bad rebase. Do you need any help resolving it @xknown?
